File: leetcode_python/Math/palindrome-number.py
# V0
class Solution(object):
    # @return a boolean
    def isPalindrome(self, x):
        length = len(str(x))
        if length==1 or length==0:
            return True
        if x < 0:
            return False
        mid = length//2 
        return str(x)[:mid] == str(x)[-mid:][::-1]

### test case
s = Solution()
assert s.isPalindrome(100) == False
assert s.isPalindrome(3456) == False
assert s.isPalindrome(1001) == True
assert s.isPalindrome(1) == True
#assert s.isPalindrome(None) == True
assert s.isPalindrome(10**32 + 1) == True
assert s.isPalindrome(10**32 + 3) == False

# V1
import math  
class Solution(object):
    # @return a boolean
    def isPalindrome(self, x):
        if x < 0:
            return False
        x = str(x)
        # if x is like 12321  (odd length)
        if str(x)[:math.floor(len(str(x))/2)] == x[math.floor(len(str(x))/2)+1:][::-1]:
            return True
        # if x is like 2222 (even length)
        if str(x)[:math.floor(len(str(x))/2)] == x[math.floor(len(str(x))/2):][::-1]:
            return True
        else:
            return False

# V2 
# Time:  O(1)
# Space: O(1)
class Solution(object):
    # @return a boolean
    def isPalindrome(self, x):
        if x < 0:
            return False
        copy, reverse = x, 0

        while copy:
            reverse *= 10
            reverse += copy % 10
            copy //= 10

        return x == reverse

File: leetcode_java/src/main/java/LeetCodeJava/Math/PalindromeNumber.java
package LeetCodeJava.Math;

// https://leetcode.com/problems/palindrome-number/

/**
 *  9. Palindrome Number
 *  Easy
 *
 *  Given an integer x, return true if x is a palindrome, and false otherwise.
 *  An integer is a palindrome when it reads the same forward and backward.
 *  For example, 121 is a palindrome while 123 is not.
 *
 *  Example 1:
 *    Input: x = 121
 *    Output: true
 *
 *  Example 2:
 *    Input: x = -121
 *    Output: false   (reads 121- from right to left)
 *
 *  Example 3:
 *    Input: x = 10
 *    Output: false
 *
 *  Constraints:
 *    -2^31 <= x <= 2^31 - 1
 *
 *  Follow up: Could you solve it without converting the integer to a string?
 */
public class PalindromeNumber {

    // V0
    // IDEA: reverse only the second half of the digits, compare with the first half
    /**
     * time = O(log x)
     * space = O(1)
     */
    public boolean isPalindrome(int x) {

        // negatives are never palindromes;
        // a number ending with 0 (except 0 itself) can't be one either
        if (x < 0 || (x % 10 == 0 && x != 0)) {
            return false;
        }

        int reverted = 0;
        while (x > reverted) {
            reverted = reverted * 10 + x % 10;
            x /= 10;
        }

        // even digit count: x == reverted
        // odd  digit count: drop the middle digit via reverted / 10
        return x == reverted || x == reverted / 10;
    }
}
